File: 力扣/每日一题/841. 钥匙和房间.py
class Solution:
    def canVisitAllRooms(self, rooms: list) -> bool:
        if len(rooms) < 1:
            return False
        room_cnt = len(rooms)
        ans_set = set()

        def dfs(n):
            if n in ans_set:
                return
            ans_set.add(n)
            if rooms[n]:
                for i in rooms[n]:
                    dfs(i)
        dfs(0)
        return True if len(ans_set) == room_cnt else False

File: c++代码/841.钥匙和房间.cpp
#include <iostream>
#include <vector>
#include <set>
using namespace std;

class Solution
{
public:
    set<int> ans_set;

    bool canVisitAllRooms(vector<vector<int>> &rooms)
    {
        int length = rooms.size();
        if (length < 1)
            return true;
        this->dfs(0, rooms);
        if (this->ans_set.size() == length)
            return true;
        else
            return false;
    }
    void dfs(int n, vector<vector<int>> &rooms)
    {
        if (this->ans_set.count(n) == 1)
            return;
        this->ans_set.insert(n);
        for (int i = 0; i < rooms[n].size(); i++)
        {
            this->dfs(rooms[n][i], rooms);
        }
    }
};

int main()
{
    return 0;
}